### ISSUE TITLE
Enhancement: Keyboard/Mouse Input, Cross Input With Controller/Keyboard

### ISSUE THREAD

**Issue by Tominounet** [KernelZechs/open-joystick-display#36]
Hey there,

First of all, I wanted to know if there were any way to use the keyboard as an input driver ?
It could be use to test mapping for example.

In my case, I wanted to use that in a very specific way. Let me explain.
I use your software for streaming/recording. Thanks by the way, you did a really great job. 

The game I play need to be on xbox one controller at a moment, and on keyboard at other moment. It would be great if I could see on my open joystick display broadcoast my stick going up as I'm holding my joystick up OR press Z (or UP) on my keyboard.

If there was a way to bind keys to input driver (so it would trigger the buttons), or to set keys on buttons, it would fit my needs I think.

Do you think this is possible ?

**Comment by KernelZechs** [comment]
Not yet but it is on the roadmap for keyboard and mouse support. I've had a few ask for it over the last year since this app went up and it may just be time to put it in.

In general though the first implementation is just going to be a keyboard/mouse driver that will simply show which keys or mouse clicks you hit. In terms of multi-mapping you could build a theme to handle that once the feature is complete.

I'll keep this ticket open and if I get a beta build I'll let you know so you can try it.

**Comment by Tominounet** [comment]
Thanks ! Really cool :) !

**Comment by JukePlz** [comment]
I think once this is on the way (in the future) having a theming feature to switch the whole image to the last used input device would be great, or a way to switch themes automatically.

When streaming, screen real state is usually at a prime and as such having a theme that shows all input devices would take a lot of space that could be saved by just showing the last input device in the same screen space. It would come in handy for games like Xanadu Next that are impossible to play with the default gamepad controls and usually requiere gamepad+keyboard+mouse for different things.

**Comment by KernelZechs** [comment]
@JukePlz it sounds like to me you want an accelerator (keyboard shortcut) to easily switch profiles while streaming.

**Comment by JukePlz** [comment]
Ideally it would be automated, accelerators can already be done by juggling scenes from the streaming program (OBS or whatever) but it involves the user duplicating the scene several times, takes extra keybinds, forces the user to touch the keyboard in games that may only need gamepad+mouse, and it's really much more cumbersome when you have to play and also keep track of the scene or viewer configuration yourself (makes you lose time in speedruns, needs extra screenspace to keep OBS or OJD in view at all times to make sure you didn't mess up, etc)

**Comment by KernelZechs** [comment]
I'm having a hard time envisioning this concept also I'm having a hard time how this is going to work cross-platform which is an absolute must for this project. What you're almost looking for is a more robust scene management tool for OBS which I believe is out of scope for this project. 

Maybe I'm misunderstanding what you're asking- maybe all you're asking is for multiple controllers on screen in OJD which is something I've been working on in OJD v3, or maybe you're asking for some sort of hook in OBS which would communicate with OJD and toggle these automatically (which leads me to a cross-platform concern, but prob. could be done with some work).

I don't want to disregard this idea entirely because I'm not 100% on what you're asking me to do. If you have time, could do me a favor and build a flow chart of your workflow process (use something like gliffy or visio- hell even photoshop or paint), something I can tangibly understand in logic, I can tell you if it is something that could be done.

**Comment by KernelZechs** [comment]
An update:

I've been fiddling around with this keyboard/mouse concept for a bit, and I'm not really liking how this functions in OJD, quite frankly the application was built with a focus on Gamepads/Joysticks/Wheels (basically static human interface devices) whereas Keyboards are very dynamic and have very particular styles, keyboard layouts (QUERTY, DVORAK, etc.), and locales (Japanese, European, etc.). 

Not to mention the type of game you're playing. MOBA players have a very particular set of keys. XIV and WoWers have a completely different set of keys from player to player. Let's not even get into RTS scene. 

If I were going to do this feature I would want it to be something that is going to work for a vast majority of users. So I don't think I'm going to move forward with this for OJD. (Also it's called Open 'Joystick' Display, doesn't seem like a good fit)

With that being said, when I get some time I might put this into another application that is generally focused on keyboards and mice since I did get the feature working as a concept. In the meantime this individual has a basic keyboard and mouse overlay that will work for your needs: https://obsproject.com/forum/resources/input-overlay.552/

If you need joysticks to mirror keyboard inputs there is software out there for that (Like JoyToKey), convoluted I know- but I'm unhappy with the outcome of the current solution and I really want to keep this application focused on its intended purpose.

**Comment by Tominounet** [comment]
Totally understand your point of view man ! I will find a way to do what I want anyway ! Thanks for the feedback :)
